drivers/ethernet/eth_sam_gmac.c

```diff
@@ -1724,7 +1724,7 @@ static enum ethernet_hw_caps eth_sam_gmac_get_capabilities(struct device *dev)
 #if GMAC_PRIORITY_QUEUE_NO >= 1
 		ETHERNET_QAV |
 #endif
-		ETHERNET_LINK_100BASE_T;
+		ETHERNET_LINK_100BASE_T | ETHERNET_TX_QUEUE;
```

Comment:
From commit message:

> However, this optimizations will not apply if the device supports
ETHERNET_TX_QUEUE where pkt will be referenced until it actually
disappear from the queue, and where more than one pkt can be stacked: so
each one requires its own ethernet header buffer.

But then name "ETHERNET_TX_QUEUE" isn't selective enough. Because the talk is a device having a queue of buffers "by reference", i.e. queue of buffers referenced by their addresses. Because almost any Ethernet controller has a queue for literal packet data (i.e. you can write next frame into controller memory before the previous one physically transmitted).

And actually, I don't even think *Ethernet controllers* have such queues, instead, it would be the problem with DMA, where DMA xfer queue exactly stores buffers by reference (to transfer literal packet data from main memory into Ethernet controller queue).

That's why eth_sam_gmac.c is the only one affected - none of the rest use DMA so far.

So, all in all, the flag needs to be renamed for clarity. Based on the above, suggested names are:

* ETHERNET_TX_QUEUE_BY_REF
* ETHERNET_TX_QUEUE_DMA

Reply:
We don't care about the technical details on what/how is the queue. It keeps a ref, that's the only important thing l2 has to know. That's all

Reply:
> We don't care about the technical details on what/how is the queue. It keeps a ref

Yes, we care only about queues which keep refs to net_buf's. As there can be many other types of "TX queue", the option should be named accordingly. Or very soon all drivers will have ETHERNET_TX_QUEUE, because, well, almost any hardware has some kind of queue.

Reply:
No L2 does not give a damn about what type of queue it is below. It just needs to know that, the pkt going to send will be queued in the driver, so it should not provide a statically allocated buffer. That's all.

Not that for hardware queues, we already have specific attributes. That's why this new attribute does not have _HW_ in its name.

Reply:
Well, let me say it again - unless that flag is named clearly, it will be just added everywhere "just in case".

Reply:
FreeBSD has AF_INET of 2, AF_INET6 of 28: https://github.com/freebsd/freebsd/blob/master/sys/sys/socket.h . Well, what can I say. Let's make more of a zoo, why not.